File: n1netails-api/src/main/java/com/n1netails/n1netails/api/service/TailStatusService.java
package com.n1netails.n1netails.api.service;

import com.n1netails.n1netails.api.exception.type.TailStatusNotFoundException;
import com.n1netails.n1netails.api.model.core.TailStatus;
import com.n1netails.n1netails.api.model.request.PageRequest;
import com.n1netails.n1netails.api.model.response.TailStatusResponse;
import org.springframework.data.domain.Page;

/**
 * Service responsible for managing tail statuses.
 *
 * <p>This service allows retrieval, creation, update, and deletion of tail statuses.</p>
 */
public interface TailStatusService {

    /**
     * Retrieves a paginated list of tail statuses.
     *
     * <p>If the search term is provided, only statuses containing the search term
     * (case-insensitive) are returned.</p>
     *
     * @param request the page request, including page number, size, sort, and optional search term
     * @return a page of tail statuses; never {@code null}, may be empty
     */
    Page<TailStatusResponse> getTailStatusList(PageRequest request);

    /**
     * Retrieves a tail status by its ID.
     *
     * @param id the identifier of the tail status
     * @return the corresponding tail status
     * @throws IllegalArgumentException if a tail status with the given ID does not exist
     */
    TailStatusResponse getTailStatusById(Long id);

    /**
     * Creates a new tail status.
     *
     * @param request the tail status data
     * @return the created tail status
     * @throws IllegalArgumentException if a tail status with the same name already exists
     */
    TailStatusResponse createTailStatus(TailStatus request);

    /**
     * Updates an existing tail status.
     *
     * @param id      the ID of the tail status to update
     * @param request the updated tail status data
     * @return the updated tail status
     * @throws IllegalArgumentException if a tail status with the given ID does not exist,
     *                                  or if another tail status already exists with the given name
     */
    TailStatusResponse updateTailStatus(Long id, TailStatus request);


    /**
     * Deletes a tail status by its ID if it is deletable.
     *
     * @param id the ID of the tail status to delete
     * @throws TailStatusNotFoundException if a tail status with the given ID does not exist
     */
    void deleteTailStatus(Long id) throws TailStatusNotFoundException;
}

File: n1netails-ui/src/main/typescript/src/app/service/tail-status.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { UiConfigService } from '../shared/ui-config.service';
import { PageRequest, PageResponse } from '../model/interface/page.interface';
import { PageUtilService } from '../shared/page-util.service';

export interface TailStatus {
  name: string;
  deletable: boolean;
}

export interface TailStatusResponse {
  id: number;
  name: string;
  deletable: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class TailStatusService {

  host: string = '';
  private apiUrl = '/ninetails/tail-status'; // Base URL for tail status operations

  constructor(
    private http: HttpClient,
    private uiConfigService: UiConfigService,
    private pageUtilService: PageUtilService
  ) {}

  createTailStatus(request: TailStatus): Observable<TailStatusResponse> {
    this.host = this.uiConfigService.getApiUrl() + this.apiUrl;
    return this.http.post<TailStatusResponse>(this.host, request);
  }

  getTailStatusList(pageRequest: PageRequest): Observable<PageResponse<TailStatusResponse>> {
    let params = this.pageUtilService.getPageRequestParams(pageRequest);
    this.host = this.uiConfigService.getApiUrl() + this.apiUrl;
    return this.http.get<PageResponse<TailStatusResponse>>(this.host, { params });
  }

  getTailStatusById(id: number): Observable<TailStatusResponse> {
    this.host = this.uiConfigService.getApiUrl() + this.apiUrl;
    return this.http.get<TailStatusResponse>(`${this.host}/${id}`);
  }

  updateTailStatus(id: number, request: TailStatus): Observable<TailStatusResponse> {
    this.host = this.uiConfigService.getApiUrl() + this.apiUrl;
    return this.http.put<TailStatusResponse>(`${this.host}/${id}`, request);
  }

  deleteTailStatus(id: number): Observable<void> {
    this.host = this.uiConfigService.getApiUrl() + this.apiUrl;
    return this.http.delete<void>(`${this.host}/${id}`);
  }
}
